Replace coreunix.Add with shell/fsnode's AddFromReader

wking opened the pull request:
coreunix.Add is basically a wrapper around BuildDagFromReader that
returns a string-ified key instead of a DAG node.  Since we'll need
the DAG node to add to the directory, converting to a key and then
looking that key back up in dirbuilder.AddChild is just extra work.

In #1127, @whyrusleeping mentioned that the file- and
directory-addition API needed some polishing, and this WIP PR is my
attempt to grind things down to a more compact API.  I'll probably
just work up to removing coreunix.Add completely, and then start a new
PR for whatever function we want to cull next.

jbenet commented:
all the text above sgtm, except:

> I'll probably just work up to removing coreunix.Add completely

why remove `coreunix.Add`? 

i want every function in the "API" to be in all, {cli, lib, http api}. `coreunix` is part of the (currently very unpolished) lib api, but it should remain to mirror things. 

the reason for mirroring the api exactly is ease of use. having a consistent api across all {cli, http api, and lib} will make using IPFS to build things very easy. we can certainly have _more_ in the lib, and guide users to use other functions, but every part of the api should exist.

I'll try to synthesize a generalized API spec next week

whyrusleeping commented:
@jbenet I gotta be straight with you, i'm not [a huge fan](http://images.t-nation.com/forum_images/1/7/177a1_ORIG-huge_fan.jpg) of the whole `coreX.Function` stuff... its a lot of thin wrappers over things that require running the daemon *in* your app to even make use of it. I think that the shell API is the way to go for making a consistent, easy to use API for users in Go

jbenet commented:
I just want the API to be implemented in the interface, shell or core doesn't really matter.



Also, everything in the shell will have to be in the core one way or another. The shell is a wrapper that can RPC over to the daemon. It has to be implemented somewhere. 





And one shell implementation may just be direct calls to the core, if you DO happen to be running an embedded node in your app. That is very much a use case. Think sqllite, leveldb, apps that build use IPFS as a database and don't want to require installing a node in the OS.




—
Sent from Mailbox

On Fri, Apr 24, 2015 at 9:38 PM, Jeromy Johnson <notifications@github.com>
wrote:

> @jbenet I gotta be straight with you, i'm not [a huge fan](http://images.t-nation.com/forum_images/1/7/177a1_ORIG-huge_fan.jpg) of the whole `coreX.Function` stuff... its a lot of thin wrappers over things that require running the daemon *in* your app to even make use of it. I think that the shell API is the way to go for making a consistent, easy to use API for users in Go
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/ipfs/go-ipfs/pull/1136#issuecomment-96128111

wking commented:
On Fri, Apr 24, 2015 at 09:24:40PM -0700, Juan Batiz-Benet wrote:
> > I'll probably just work up to removing coreunix.Add completely
> 
> why remove `coreunix.Add`? 
> 
> i want every function in the "API" to be in all, {cli, lib, http
> api}. `coreunix` is part of the (currently very unpolished) lib api,
> but it should remain to mirror things.
> 
> the reason for mirroring the api exactly is ease of use. having a
> consistent api across all {cli, http api, and lib} will make using
> IPFS to build things very easy. we can certainly have _more_ in the
> lib, and guide users to use other functions, but every part of the
> api should exist.

And I'm fine with that.  It just seems to me that *importer* is
currently where reader → file-node lives.  I don't want to mirror that
or write a big stack of wrappers on top of it.  We currently use
BuildDagFromReader in coreunix.Add(), coreunix.add(), and
commands.add().  I want:

  lib location → use in command-line or HTTP-endpoint implementation

not the current:

  lib location → coreunix wrapper → command wrapper → use in
    command-line or HTTP-endpoint implementation

If you want the lib location to be in coreunix, or lib/ipfs/unix or
wherever and just want the chunking algorithms in ‘importer’, that's
fine with me.  But I think BuildDagFromReader is closer to the …err…
building-a-DAG-node-from-a-reader signature than the current
collection of add wrappers ;).

> I'll try to synthesize a generalized API spec next week

Sounds good.  In the interim, I think removing any extra layers of
thin wrappers makes sense, and it will make it easier to move things
over to the generalized API once we've decided what that will look
like ;).

whyrusleeping commented:
> I think removing any extra layers of thin wrappers makes sense
:+1:

jbenet commented on util/testutil/addcat/addcat.go:
```diff
@@ -0,0 +1,44 @@
+	errors "github.com/ipfs/go-ipfs/util/debugerror"
+)
+
+func AddCat(adder *core.IpfsNode, catter *core.IpfsNode, data string) error {
```
The creation of this function is a good addition, let's keep this.

jbenet commented on core/corehttp/gateway_test.go:
```diff
@@ -65,10 +66,19 @@ func newNodeWithMockNamesys(t *testing.T, ns mockNamesys) *core.IpfsNode {
+		strings.NewReader("fnord"),
+		n.DAG,
+		n.Pinning.GetManual(),
+		chunk.DefaultSplitter)
```
:-1: I'm not ok with this change. This forces all those locations to understand what (a) a dagservice is, (b) what pinning is, (c) what chunking is. This is moving up a ton of complexity in places it does not belong. The ability to sink down and do this is good to have and should continue to be there, but _removing a useful_ abstraction is not good. Particularly when users of ipfs as a library will be looking at our code and following suit.

jbenet commented:
I very much appreciate the effort to remove unnecessary code, and unnecessary symbols. In general, I'm all for that. And thanks for taking the time to look through this case.

But in this case, I'm strongly against this change. It _increases_ (not _reduces_) the complexity for us, and for end users. When i want to add an `io.Reader` to my node, i really don't want to _have to_ know what a dagservice, pinning, and chunking all are.  (of course, i may _need to_, if i want to do something fancy, like use my own chunker, but _sane defaults_ should be _trivial_ to use). This removal also increases the potential for errors / bug surface area.

I'm ok with the `testutil.AddCat` abstraction, as that removes boilerplate test code and reduces the surface area of what could go wrong.

jbenet commented:
I know this is not documented anywhere, please bear with me as I make all this clearer through specs.

I don't really care whether it's `MyFunc(nd, ...)` or `ndinterface.MyFunc(...)`. What i care about is there being a standard interface for everything a node does, as _simple_ as possible for the end user. this is a UX thing. these thin wrappers here are really beneficial to end users who do not know much about how the system works, and do not want to sink knee-deep into it to be able to use it. `ipfs add file` is enough in the terminal. so should something like `ipfs.Add(reader)` in Go.

For now (and until the shell interface appears), please look at the `core*` functions as a (messy) interface that defines _the entire_ functionality of an IPFS node, in Go. An end user who embeds IPFS into her application should be able to use all the `core*` functions which will be familiar from the cli + api. 

@whyrusleeping 

> its a lot of thin wrappers over things that require running the daemon in your app to even make use of it

Yeah, that's fine.  the ability to embed an IPFS node and use it directly from within an application (as you would an HTTP client or server) is very much a target use case. Not all IPFS nodes will be bulky, os-level services. Many will be ephemeral nodes created inside apps. (and yeah, the shell interface -- when it arrives -- will allow embedding only a proxy object and RPCing out to use it. ideally we will use the shell interface for the directly embedded case too, so the interface is the exact same, but until then, `core*` is what we have)

wking commented:
On Mon, Apr 27, 2015 at 02:29:16AM -0700, Juan Batiz-Benet wrote:
> What i care about is there being a standard interface for everything
> a node does, as _simple_ as possible for the end user. this is a UX
> thing.

I agree that having a simple UX is really important.  I just think
having a single interface to perform a task (e.g. creating a file node
from a reader) is going to be simpler than having a range of similar
functions for performing that task with varying numbers of knobs.
Reading a brief docs saying “if you don't have a specific chunker in
mind, use chunk.DefaultSplitter()” and similarly for the other options
seems easier from a UX perspective than “select the reader-based-adder
from this list that best matches your desired signature”.

> these thin wrappers here are really beneficial to end users who do
> not know much about how the system works, and do not want to sink
> knee-deep into it to be able to use it. `ipfs add file` is enough in
> the terminal. so should something like `ipfs.Add(reader)` in Go.

Folks using a terminal or an HTTP API will have a different interface
because there's a limit to what you can fit down a socket (e.g. data
is easy, methods are not).  I agree that the Go API should be
*familiar* to folks who already know the command line or HTTP
interface, but I think forcing an exact match (e.g. returning the key
string from Add instead of the file node from which the key string can
be extracted, but which also supports other actions) just makes the
library interface more complicated.  Because some folks *will* need
access to the file node, and that means either a parallel function
with a slightly different syntax, or they'll need a workaround like
dirbuilder.AddChild to recover the node from the key (see d77b90c1).

Also, adding a file from a reader is going to get more complicated as
the amount of per-file metadata we track increases (access modes,
owner and group IDs, modification times?, …).  I think you can make a
much stronger case for the current coreunix.AddR than for the current
coreunix.Add.  Adding files *from a reader* (vs. from a path) seems
peripheral/advanced enough that it doesn't need an entry-level
wrapper, even if you feel like entry-level wrappers are a useful tool.

Anyhow, I think we have the same end goal in mind (providing an
accessible API to library consumers), we just have different views on
what would be most accessible.  If we have the dev time and clear
separation, I'm ok with having parallel low- and high-level APIs, but
until we have a clear and stable low-level API I don't think it's
worth maintaining sugar-coated wrappers on top, when a few lines of
documentation seem to address your “how do I use this” concerns ;).
But if you're not buying what I'm selling, I'm happy to rebase that
change to the end of my branch :p.

jbenet commented:
> having a single interface to perform a task (e.g. creating a file node
from a reader) is going to be simpler than having a range of similar
functions for performing that task with varying numbers of knobs.

agreed, which is why i prefer a simple:

```
Add(io.Reader)
```
which will cover most of users' first steps into using this.

> I think forcing an exact match (e.g. returning the key
string from Add instead of the file node from which the key string can
be extracted, but which also supports other actions) just makes the
library interface more complicated.

changing the signature from: 

```go
// from
func Add(...) Key

// to
func Add(...) Node
```

is fine with me. what isn't fine is adding lots of knowledge to the user (dagservice, pinning, chunker).

We already ask users to understand a lot of stuff in the model. complicating it even further with more implemenetation-specific knowledge is not good.


> Also, adding a file from a reader is going to get more complicated as
> the amount of per-file metadata we track increases (access modes,
> owner and group IDs, modification times?, …). 

That's fine, all of those can be more advanced function calls that do ask the user to understand them. What i'm rejecting here is _removing_ the simplest possible use case, which most users will fall into at the very beginning.

> Adding files *from a reader* (vs. from a path) seems
> peripheral/advanced enough that it doesn't need an entry-level
> wrapper, even if you feel like entry-level wrappers are a useful tool.

Not really, this is what I use, and what other people use when they first try to add data with Go. It's very natural -- both from an idiomatic Go perspective (manipulating io.Readers) and from what we established with the cli (`ipfs add <file>`)

>  If we have the dev time and clear separation, I'm ok with having parallel low- and high-level APIs, but until we have a clear and stable low-level API I don't think it's worth maintaining sugar-coated wrappers on top

It takes _waaay_ more time explaining to users on irc what the more complicated things mean, and how to use them. and those are just the users brave enough to ask. most give up on the spot, and i've seen it first hand. At this point i've run multiple "classes" with 30+ people hacking with ipfs. I've observed people trying to grasp the large number of concepts when their goal is a simple "i want to add this data into ipfs".

I'm interested in dropping the barriers of entry as much as possible, and this requires having _extremely_ simple interfaces that do the simplest possible thing -- the average use case -- by default.  Advanced use should be optional, not forced.

> when a few lines of documentation seem to address your “how do I use this” concerns

The problem is not 5 sentences of doc. the problem is 5 sentences here, 5 sentences there, and very soon we're at 1000s of sentences of doc. we already ask waaaay too much from users. ipfs must be extremely simple to use, and leveraging the analogues we establish in the cli is critical to this. 

take a look at Go's net and http packages. they hide away a _ton_ of complexity for the common case. the more advanced use cases can do things by dropping down into lower levels. crystal clear, unencumbered interfaces are really valuable to users new to IPFS, new to Go, new to content addressing in general.... Compare this to asking users to enter:

```
ipfs add <file>
ipfs add <dagservice> <pinner> <splitter> <file>
```

wking commented:
On Mon, Apr 27, 2015 at 01:53:49PM -0700, Juan Batiz-Benet wrote:
> > Adding files *from a reader* (vs. from a path) seems
> > peripheral/advanced enough that it doesn't need an entry-level
> > wrapper, even if you feel like entry-level wrappers are a useful
> > tool.
> 
> Not really, this is what I use, and what other people use when they
> first try to add data with Go. It's very natural -- both from an
> idiomatic Go perspective (manipulating io.Readers) and from what we
> established with the cli (`ipfs add <file>`)

No, that's adding a file from a path (where coreunix.AddR is what you
want).  I like that API a lot more, because paths also give you access
to the access mode, owner ID, ….  What coreunix.Add is like is adding
from stdin:

  $ ipfs add </path/to/file

and we *don't* have a command-line interface to do that.

> I'm interested in dropping the barriers of entry as much as
> possible, and this requires having _extremely_ simple interfaces
> that do the simplest possible thing -- the average use case -- by
> default.  Advanced use should be optional, not forced.

Ok, and on IRC you suggested to levels of API:

* core = kernel. it defines that an IPFS node is.
* shell = a wrapper around the core that makes the api very simple +
  clean for consumers, and also makes it possible to use across rpc.

So having high-level wrappers around adding and accessing files in
shell/fsnode.go works for me.  And maybe moving the lower-level
BuildDagFromReader to AddFromReader core/fsnode would fill the
low-level API slot.

> take a look at Go's net and http packages. they hide away a _ton_ of
> complexity for the common case. the more advanced use cases can do
> things by dropping down into lower levels. crystal clear,
> unencumbered interfaces are really valuable to users new to IPFS,
> new to Go, new to content addressing in general.... Compare this to
> asking users to enter:
> 
> ```
> ipfs add <file>
> ipfs add <dagservice> <pinner> <splitter> <file>

The command line is a lot easier here, because you can have optional
arguments with sane defaults (not possible without hoop-jumping in a
single Go function).  I'm +1 on:

  $ ipfs add [options] <file>

with options like:

  --dagservice   address of the IPFS node to use.  Defaults to
                 manipulating your local ~/.ipfs directory directly.
  --pinner       pinner to use.  Select from ….  Defaults to
                 recursively pinning each added object.
  --splitter     algorithm for chunking files into content-addressable
                 blocks.  Select from 'rabin' or 'size'.  Defaults to
                 'size'.  For details on the options, use:

                   $ godoc github.com/ipfs/go-ipfs/importer/chunk

Of course, the command line UI is going to be less flexible about
output types (you can't just return a *dag.Node), but it is nice to
have the flexibility of options that newcomers can ignore until they
really *need* to use Rabin chunking (or whatever).  There's no need
for separate high-level and low-level interfaces, but you may want to
separate your options into high-level and low-level options.

jbenet commented:
> `ipfs add </path/to/file`

sure

> and we *don't* have a command-line interface to do that.

what do you mean? we do. `ipfs add </path/to/file` works just fine. try it:

```
> echo "hello wking" | ipfs add
added QmciRKQJNeejQ8qT8mNgCuDXGmY7VfMfREHWDFyu6M8aK9
```

> not possible without hoop-jumping in a single Go function)

it's ok to have multiple functions. i prefer this. but the _simplest, most straightforward_ case that users will want to try first should be trivially easy to spot and use. hence `Add(io.Reader)`.

wking commented:
On Mon, Apr 27, 2015 at 03:58:27PM -0700, Juan Batiz-Benet wrote:
> > and we *don't* have a command-line interface to do that.
> 
> what do you mean? we do. `ipfs add </path/to/file` works just
> fine. try it:
> 
> ```
> > echo "hello wking" | ipfs add
> added QmciRKQJNeejQ8qT8mNgCuDXGmY7VfMfREHWDFyu6M8aK9
> ```

Ahh, good stuff.  Docs updated in #1151.  I guess we can pull the
intended user/group from the process's effective user/group and umask,
so that's not too bad (but does mean the local client has to do more
legwork to setup a call to an IPFS daemon).

jbenet commented:
> we can pull the intended user/group from the process's effective user/group and umask, so that's not too bad (but does mean the local client has to do more legwork to setup a call to an IPFS daemon).

yeah, sounds good. 

i wonder if this should be an option. for example, most cases when publishing to the web, i do not want ipfs to grab my user or permissions, only the content. something like `--keep-permissions` or something.

jbenet commented:
(though it doesnt matter much either way)

wking commented:
On Mon, Apr 27, 2015 at 04:33:23PM -0700, Juan Batiz-Benet wrote:
> i wonder if this should be an option. for example, most cases when
> publishing to the web, i do not want ipfs to grab my user or
> permissions, only the content. something like `--keep-permissions`
> or something.

So “please don't wrap this file with a metadata node”, which makes it
easy to create the file, and could be interpreted on unpacking as “use
the current user/group, consult the current umask for the access mode,
use the current time for atime/mtime, …).

Do you want that boolean flag to be part of the high-level
add-from-reader signature?

  func Add(node *core.IpfsNode, reader io.Reader, metadata bool) (*dag.Node, err)

where ‘metadata == true’ means “wrap the top-level file node with
mtetadata based on the current environment” and ‘metadata == false’
means “just return the top-level file node”.

Or should the high-level signature just assume we don't want to guess
at metadata and use unwrapped files?

jbenet commented:
> Or should the high-level signature just assume we don't want to guess at metadata and use unwrapped files?

i think this is right. we could have a set of functions that use metadata. it's an additional concept people need to know and want. its possible it'll be common enough to warrant being the default, but im not convinced of that yet.

wking commented:
I'm still running through the local tests, but I think this is ready
to land.  I've shifted coreunix.Add to shell/fsnode's AddFromReader
and return a dag.Node there instead of a stringified key.  I've also
added a deprecation warning to coreunix.Add following #1159.

core/corehttp/gateway_test.go would be a bit simpler if it could use
fsnode.AddFromReader instead of importer.BuildDagFromReader [1], but
we can't do that without either externalizing the test from the
corehttp package or breaking our anti-cyclic-import hierarchy (#1158).
Personally, using importer.BuildDagFromReader directly from this test
seems like the simplest choice, but I'm happy to externalize the test
if that's the only thing holding up this PR.

[1]: https://github.com/ipfs/go-ipfs/pull/1136/files#r29130459

jbenet commented on CHANGELOG.md:
```diff
@@ -1,5 +1,36 @@
+
+That's a bit more verbose if all you want is the stringified key, but
+returning a `dag.Node` makes it easier to perform other operations
+like adding the file node to a directory node.
```
we're not currently doing descriptions like this. seems very nice and useful for consumers.

the binary's changelog is not the best place. this changelog is just a tiny summary. maybe we can have a different file for things like this.

we should define clearly what interfaces we commit to as stable. in particular:

- core
- shell

come to mind as the main ones. changes in these are potentially large.

wking replied:
On Wed, Apr 29, 2015 at 01:26:36PM -0700, Juan Batiz-Benet wrote:
> the binary's changelog is not the best place. this changelog is just
> a tiny summary. maybe we can have a different file for things like
> this.

Git uses a directory of per-release notes [1].  I'm happy to put this
in docs/release-notes/unreleased.md (to be moved to
docs/release-notes/0.3.4.md when you cut that release) if that sounds
reasonable to you.

[1]: http://git.kernel.org/cgit/git/git.git/tree/Documentation/RelNotes?id=v2.3.7

jbenet replied:
sure, sounds good.

jbenet commented on cmd/ipfs/init.go:
```diff
@@ -132,23 +133,24 @@ func addDefaultAssets(out io.Writer, repoRoot string) error {
 	defer nd.Close()
 
-	dirb := uio.NewDirectory(nd.DAG)
+	dir := &merkledag.Node{Data: unixfs.FolderPBData()}
```
Curious why not using the constructor?

wking replied:
On Wed, Apr 29, 2015 at 01:27:43PM -0700, Juan Batiz-Benet wrote:
> > @@ -132,23 +133,24 @@ func addDefaultAssets(out io.Writer, repoRoot string) error {
> >  	}
> >  	defer nd.Close()
> >  
> > -	dirb := uio.NewDirectory(nd.DAG)
> > +	dir := &merkledag.Node{Data: unixfs.FolderPBData()}
> 
> Curious why not using the constructor? 

Because I wanted AddNodeLink.  Cycling from node → key string → node →
AddNodeLink (which is what AddChild does [1]) seemed wasteful and
awkward (see note in 0ce5784a, cmd/ipfs/init: Replace coreunix.Add
with importer.BuildDagFromReader, 2015-04-24).

[1]: https://github.com/ipfs/go-ipfs/blob/cc5f6bb306430d146d3e5a3f7956073c10d0112b/unixfs/io/dirbuilder.go#L30

rht replied:
or define a method `dirb.AddNodeLink(node)`.

jbenet commented on shell/fsnode/fsnode.go:
```diff
@@ -0,0 +1,10 @@
+* extract file and directory nodes to your local filesytem, and
+* print file contents to writers.
+*/
+package fsnode
```
i think this package should be called `unixfs` as that's the abstraction that this is calling to.

wking replied:
On Wed, Apr 29, 2015 at 01:30:04PM -0700, Juan Batiz-Benet wrote:
> > +package fsnode
> 
> i think this package should be called `unixfs` as that's the
> abstraction that this is calling to.

Will do, although I don't see what's especially Unix-y about these
file nodes.

wking replied:
On Wed, Apr 29, 2015 at 01:56:43PM -0700, W. Trevor King wrote:
> Wed, Apr 29, 2015 at 01:30:04PM -0700, Juan Batiz-Benet:
> > > +package fsnode
> > 
> > i think this package should be called `unixfs` as that's the
> > abstraction that this is calling to.
> 
> Will do, although I don't see what's especially Unix-y about these
> file nodes.

Done, although I still prefer fsnode or posixfs to unixfs.

jbenet commented on core/coreunix/add.go:
```diff
@@ -22,8 +22,12 @@ import (
+// node.
+//
+// This function is deprecated and will be removed in 0.4.0.  You
+// should use shell/fsnode's AddFromReader instead.
```
@wking sorry--  i don't think we're communicating...  i don't want this deprecated from core. i don't see the why it needs to be yet.

I'm not 100% sure about this right now, but it seems to me that all the things that the shell can do, will likely be in core as well. (the shell is mostly just a wrapper). Perhaps I should spend time charting out what shell and core should look like, i can do that next sprint.

the core is where the bulk of the operations will remain, the shell is just a very simplified interface that makes it easier for users to use ipfs nodes that are either embedded or over an rpc transport.

okay, i haven't made things very clear with all this. we should spec out both the shell and the core should be -- on their own as interface specs -- before deciding what to deprecate, or actually deprecating, from the current interfaces. trying to decide one thing goes here and not there without the full picture -- and with clearly different mental models -- is going to cause undue decision conflicts. i'll make it a priority to get that out next week.

wking replied:
On Wed, Apr 29, 2015 at 01:41:12PM -0700, Juan Batiz-Benet wrote:
> I'm not 100% sure about this right now, but it seems to me that all
> the things that the shell can do, will likely be in core as
> well. (the shell is mostly just a wrapper). Perhaps I should spend
> time charting out what shell and core should look like, i can do
> that next sprint.

I'd like to shift importer.BuildDagFromReader into the core (as
core/unixfs's AddFromReader?  In any case it should match what's
currently my shell/fsnode AddFromReader, with shell → core).  That
seems like it has all of our current bells and whistles for creating a
DAG node from a reader, so it's a good low-level interface.

> the core is where the bulk of the operations will remain, the shell
> is just a very simplified interface that makes it easier for users
> to use ipfs nodes that are either embedded or over an rpc transport.

Exactly, I just think the function known as BuildDagFromReader *is*
the low-level interface (although I don't mind if you want to shuffle
its signature around when we move it into core).  If someone wants the
high-level interface, they can use shell/unixfs (or shell/fsnode, or
whatever we call it).  But I see no reason for a complete copy of the
high-level API into the low-level core space.

> okay, i haven't made things very clear with all this. we should spec
> out both the shell and the core should be -- on their own as
> interface specs -- before deciding what to deprecate, or actually
> deprecating, from the current interfaces. trying to decide one thing
> goes here and not there without the full picture -- and with clearly
> different mental models -- is going to cause undue decision
> conflicts. i'll make it a priority to get that out next week.

I'm fine with this, but I'm also fine deciding now that the current
coreunix.Reader is *not* going to be an appropriate low-level
interface (it's missing at least a knob to select your desired
chunker).  Anyhow, we can certainly put off the deprecation (and maybe
this whole PR?) until you've had time to think this over, but I think
it will be easier to evolve towards a stable API spec than it will be
to design it all at once before focusing on implementation.

jbenet commented:
> I've shifted coreunix.Add to shell/fsnode's AddFromReader and return a dag.Node there instead of a stringified key.  

ok

> I've also added a deprecation warning to coreunix.Add following #1159.

i think we need to properly spec out these interfaces before deciding whether anything should be deprecated. see comment in: https://github.com/ipfs/go-ipfs/pull/1136#discussion_r29378469

> but we can't do that without either externalizing the test from the corehttp package or breaking our anti-cyclic-import hierarchy (#1158).

why not using it as a separate package, like  `corehttp_test`? maybe i'm missing something.

wking commented:
On Wed, Apr 29, 2015 at 01:45:48PM -0700, Juan Batiz-Benet wrote:
> > but we can't do that without either externalizing the test from
> > the corehttp package or breaking our anti-cyclic-import hierarchy
> > (#1158).
>
> why not using it as a separate package, like `corehttp_test`? maybe
> i'm missing something.

Right, that's “extenalizing the test” ([1] and comments in 87001b78,
util/testutil/addcat/addcat: Factor out add/cat/verify helper,
2015-04-26).

[1]: http://dave.cheney.net/2014/12/01/five-suggestions-for-setting-up-a-go-project

wking commented:
On Wed, Apr 29, 2015 at 01:59:07PM -0700, W. Trevor King wrote:
> On Wed, Apr 29, 2015 at 01:45:48PM -0700, Juan Batiz-Benet wrote:
> > > but we can't do that without either externalizing the test from
> > > the corehttp package or breaking our anti-cyclic-import
> > > hierarchy (#1158).
> >
> > why not using it as a separate package, like `corehttp_test`?
> > maybe i'm missing something.
> 
> Right, that's “extenalizing the test” ([1] and comments in 87001b78,
> util/testutil/addcat/addcat: Factor out add/cat/verify helper,
> 2015-04-26).

Done in 390ed4be (core/corehttp/gateway_test: Externalize test to use
shell/unixfs' AddFromReader, 2015-04-30), although I still prefer
avoiding the high-level shell/… API when testing the low-level core/…
API.  I put a few more reasons in the commit message, and will happily
drop that commit if you find them convincing enough :p.

Anyhow, I think that addresses all of your concerns, with the
exception of “i don't want [coreunix.Add] deprecated from core” [1],
where my position is “I'm fine with [waiting for you to spec an API],
but I'm also fine deciding now that the current coreunix.Reader is
*not* going to be an appropriate low-level interface (it's missing at
least a knob to select your desired chunker)” [2].

[1]: https://github.com/ipfs/go-ipfs/pull/1136#issuecomment-97580638
[2]: https://github.com/ipfs/go-ipfs/pull/1136#issuecomment-97580638

jbenet commented on util/testutil/addcat/addcat.go:
```diff
@@ -0,0 +1,39 @@
+	errors "github.com/ipfs/go-ipfs/util/debugerror"
+)
+
+func AddCat(adder *core.IpfsNode, catter *core.IpfsNode, data string) error {
```
why a `string` and not `[]byte`? 

it's a test so it doesn't matter much, but this likely should be `[]byte`. we want to deal with large blocks of data as `[]byte`.

jbenet replied:
yeah, i dont think the change to string is useful. and it drops out a bunch of unnecessary changes.

wking replied:
On Fri, May 01, 2015 at 04:56:06AM -0700, Juan Batiz-Benet wrote:
> > +func AddCat(adder *core.IpfsNode, catter *core.IpfsNode, data string) error {
> 
> why a `string` and not `[]byte`? 

From the commit message for 87001b78 (util/testutil/addcat/addcat:
Factor out add/cat/verify helper, 2015-04-26):

  I used a string (instead of []byte) for AddCat's data, because we
  don't need to mutate the value, and Go's strings are just immutable
  byte arrays…

Also related are my comments from 3af39ac3 (cmd/ipfs/init:
bytes.NewBufferString -> strings.NewReader, 2015-04-25):

  From the docs for strings.NewReader [1]:

  It is similar to bytes.NewBufferString but more efficient and
  read-only.

  [1]: https://golang.org/pkg/strings/#NewReader

Basically, I agree with Joshua Poehls [2]:

  Ask not when to use string but rather, when to use []byte. Always
  start with string and switch to []byte when justified.

> it's a test so it doesn't matter much, but this likely should be
> `[]byte`. we want to deal with large blocks of data as `[]byte`.

Do we need the mutability in many places?  If so, I agree that it's
probably better to use []byte everywhere.  But otherwise it seems
better to use the immutable type as widely as possible and only
perform a copy-conversion to a mutable type when we need the
mutability.

[2]: http://joshua.poehls.me/2014/04/go-101-string-or-byte-slice/

whyrusleeping replied:
I think we will want to use a `[]byte` here simply because so many other potential callers of this will generate data in `[]byte` form, and turning that into a `string` type requires a full copy of the entire data. For larger `AddCat` tests, this may be copying hundreds of MBs of data, not something i'm a fan of

wking replied:
On Sat, May 02, 2015 at 02:43:28AM -0700, Jeromy Johnson wrote:
> > +func AddCat(adder *core.IpfsNode, catter *core.IpfsNode, data string) error {
> 
> I think we will want to use a `[]byte` here simply because so many
> other potential callers of this will generate data in `[]byte` form,
> and turning that into a `string` type requires a full copy of the
> entire data. For larger `AddCat` tests, this may be copying hundreds
> of MBs of data, not something i'm a fan of

This cuts the other way too if our potential callers are generating
strings and need to convert them to []bytes.  Why are the majority of
our callers generating mutable data?

whyrusleeping replied:
because the most common interface for moving data around in go is the `io.Reader`, and it reads into a `[]byte` buffer from some source (in our data generation case, an RNG)

wking replied:
On Sat, May 02, 2015 at 11:20:37AM -0700, Jeromy Johnson wrote:
> because the most common interface for moving data around in go is
> the `io.Reader`, and it reads into a `[]byte` buffer from some
> source (in our data generation case, an RNG)

Sold.  Rerolled using []byte instead of string for AddCat's data.

jbenet commented on core/coreunix/add.go:
```diff
@@ -22,8 +22,12 @@ import (
+// node.
+//
+// This function is deprecated and will be removed in 0.4.0.  You
+// should use shell/unixfs's AddFromReader instead.
```
please remove this comment. regardless of the future, today, this **is not** deprecated.

wking replied:
On Fri, May 01, 2015 at 04:58:19AM -0700, Juan Batiz-Benet wrote:
> > -// Add builds a merkledag from the a reader, pinning all objects to the local
> > -// datastore. Returns a key representing the root node.
> > +// DEPRECATED: Add builds a merkledag from the a reader, pinning all
> > +// objects to the local datastore. Returns a key representing the root
> > +// node.
> > +//
> > +// This function is deprecated and will be removed in 0.4.0.  You
> > +// should use shell/unixfs's AddFromReader instead.
> 
> please remove this comment. regardless of the future, today, this
> **is not** deprecated.

Most of my later commits in this branch (and some of the earlier ones)
are responding to this deprecation.  I've shuffled things around a
bit, so I think you should be able to merge the following initial
commits without controversy:

* 0ce5784 cmd/ipfs/init: Replace coreunix.Add with importer.BuildDagFromReader
* bf9b9fc cmd/ipfswatch/main: Replace coreunix.Add with coreunix.AddR
* ac0a606 cmd/ipfs/init: bytes.NewBufferString -> strings.NewReader

Once we've resolved strings vs []bytes [1], we can merge:

* 7f62410 util/testutil/addcat/addcat: Factor out add/cat/verify helper
* ac0a606 cmd/ipfs/init: bytes.NewBufferString -> strings.NewReader

And then once we've decided to deprecate coreunix.Add we can merge:

* cd58a84 shell/unixfs: Add a high-level filesystem-node package
* 56807e9 core/coreunix/add: Deprecate Add
* 2409fee util/testutil/addcat: importer.BuildDagFromReader -> unixfs.AddFromReader
* 159783f test/supernode_client/main.go: Migrate to unixfs.AddFromReader
* 5d97c8c core/corehttp/gateway_test: Replace coreunix.Add with importer.BuildDagFromReader
* 8a4b861 core/corehttp/gateway_test: Externalize test to use shell/unixfs' AddFromReader

Until we want to deprecate coreunix.Add, the shell/unixfs stuff is
just going to add another interface to maintain (instead of migrating
us to a new interface), so I think it's just going to cause confusion.
There's no rush though, so I'm happy to just let those commits cook
until we are ready to axe coreunix.Add.

If that sounds reasonable, I'm fine with you just merging up to the
point you're comfortable with from this branch.  I'm also happy
opening new PRs for the non-controversial changes, or pushing the
controversial changes out into new PRs.  Whatever works best for you.

[1]: https://github.com/ipfs/go-ipfs/pull/1136/files#r29509838

jbenet commented:
To merge: 

- [ ] remove deprecation statement. the function **is not** deprecated at this time.
- [ ] switch `string` to `[]byte` in `AddCat` test util
- [ ] rebase + squash

wking commented:
On Fri, May 01, 2015 at 05:09:53AM -0700, Juan Batiz-Benet wrote:
> - [ ] remove deprecation statement. the function **is not**
>   deprecated at this time.

I'm fine delaying the commits relevant to this change.  More
discussion in [1].

> - [ ] switch `string` to `[]byte` in `AddCat` test util

I'd like more motivation for []byte before using a mutable type.  More
discussion in [2].

> - [ ] rebase + squash

Happy to reroll to squash whatever you want squashed, but I've been
trying hard to write orthogonal, atomic commits that don't break
tests.

[1]: https://github.com/ipfs/go-ipfs/pull/1136/files#r29510887
[2]: https://github.com/ipfs/go-ipfs/pull/1136/files#r29509838

whyrusleeping commented:
> Happy to reroll to squash whatever you want squashed, but I've been
trying hard to write orthogonal, atomic commits that don't break
tests.

I'm this way too, but I think @jbenet prefers fewer commits per PR. The reasoning (as far as I know) is that it makes it easier to bisect if things go wrong later, fewer commits means fewer hops through the code to find where something broke. So ive gone from my usual habit of 'commit super frequently because commits are cheap and awesome' to more of a 'this looks like a decent changeset, i might commit' attitude.

wking commented:
On Sat, May 02, 2015 at 02:49:31AM -0700, Jeromy Johnson wrote:
> > Happy to reroll to squash whatever you want squashed, but I've
> > been trying hard to write orthogonal, atomic commits that don't
> > break tests.
> 
> I'm this way too, but I think @jbenet prefers fewer commits per
> PR. The reasoning (as far as I know) is that it makes it easier to
> bisect if things go wrong later, fewer commits means fewer hops
> through the code to find where something broke.

‘git bisect …’ can be fully automated, but once you find the breaking
commit you need to read through it to find the bug by hand.  That
seems like an argument in favor of smaller, atomic commits to me.

jbenet commented:
on the small commits, i should clarify: my only hard requirement is that:
- commits pass all the tests.

softer requirement is:
- each commit is an atomic commits.

@wking's atomic commits is precisely what i'd shoot for. the only hard requirements is that **_all the tests pass on every commit_**.

wking commented:
On Sat, May 02, 2015 at 12:25:51PM -0700, Juan Batiz-Benet wrote:
> the only hard requirements is that **_all the tests pass on every commit_**.

Agreed, and I've been running ‘make test_all_commits’ occasionally
(but it takes a while ;).  I'll report back on how the current e352e32
branch does once that finishes.

whyrusleeping commented:
@wking whats the status here?

whyrusleeping commented:
closing, old PR cleanup time.
